cqf-fhir-cr-hapi/src/main/java/org/opencds/cqf/fhir/cr/hapi/cdshooks/discovery/CrDiscoveryService.java

```diff
@@ -144,17 +144,39 @@ protected StringBuilder getCodesStringBuilder(List<String> ret, StringBuilder co
     }
 
     public List<String> createRequestUrl(IDataRequirementAdapter dataRequirement) {
-        if (dataRequirement == null || !dataRequirement.hasType() || !isPatientCompartment(dataRequirement.getType()))
+        if (dataRequirement == null) {
             return List.of();
+        }
+
+        // if we have a fhirQueryPattern extensions, use them
+        var fhirQueryExtList = dataRequirement.getExtension().stream()
+                .filter(e -> e.getUrl().equals(CQF_FHIR_QUERY_PATTERN) && e.getValue() != null)
+                .toList();
+        if (!fhirQueryExtList.isEmpty()) {
+            List<String> urlList = new ArrayList<>();
+            for (var fhirQueryExt : fhirQueryExtList) {
+                urlList.add(((IPrimitiveType<?>) fhirQueryExt.getValue()).getValueAsString());
+            }
+            return urlList;
+        }
+
+        // else build the query
+        if (!dataRequirement.hasType() || !isPatientCompartment(dataRequirement.getType())) {
+            return List.of();
+        }
         var patientRelatedResource = dataRequirement.getType() + "?"
                 + getPatientSearchParam(dataRequirement.getType())
                 + "=Patient/" + PATIENT_ID_CONTEXT;
+
+        // TODO: Add valueFilter extension resolution
+        // http://hl7.org/fhir/extensions/5.1.0/StructureDefinition-cqf-valueFilter.html
+
         if (dataRequirement.hasCodeFilter()) {
             return createRequestUrlHasFilters(dataRequirement, patientRelatedResource);
         } else {
-            List<String> ret = new ArrayList<>();
-            ret.add(patientRelatedResource);
-            return ret;
+            List<String> urlList = new ArrayList<>();
```

Comment:
Does this List need to be mutable?  if so, why not return List.of(patientRelatedResource)?